### ISSUE TITLE
How to contribute

### ISSUE THREAD

**Issue by jdbranham** [StephenOTT/Quintessential-Tasklist-Zeebe#31]
I'm really interested in your project.
I've been working on something similar based on the zeebe simple-tasklist, but your design is much better.
How far along is this project? Is it a serious project, or were you just prototyping?
I'd love to contribute but I'm not where to begin - I attempted to do `mvn test` but it just seems to hang waiting for some external connections it appears.

**Comment by StephenOTT** [comment]
The first phase was to scope the components out and get understanding of the different unknowns.  That was done.  
Current code works as a long polling zeebe client that can collect user task jobs and create the entities in the DB and then have a user complete the task using a formio form and it will validate the submission, followed by completion of zeebe job. 

First phase was built in vertx. But with updates to micronaut and their updates for reflection free aspect annotations(Such as their bean validator), I will be migrating the code to micronaut and kotlin. Want to reduce a bunch of boilerplate code that micronaut and kotlin will take care of. 

The migration is just a code migration, it's all the same logic that's already in code.  Hoping to get back to it in the next few weeks.  

The test is just a placeholder to generate the Mongo db and zeebe engines.  

What sort of use case do you have for the user tasks?

**Comment by StephenOTT** [comment]
Thanks for the details. That helps a lot to understand.  

So from your description I am seeing many cases of long running processes (processes that could be alive for hours, days, weeks, months etc). 

Take note of this limitation in zeebe at the moment https://github.com/zeebe-io/zeebe/issues/3254. It prevents the ability to migrate a active process instance to a new version. 

For your use case the above i think will end up being a problem because users will only be able to modify workflows for the next created instance. But won't be able to modify active instances.  Example: you generate a task to review and approve a doc/node/page by the end of the month.  If you decide 1 week into the month that you need to update the workflow, you can't effect the already active instances. 


Another consideration for you guys is the level of users you are expecting to model the bpmns and manage the various configurations: it generally takes a skilled user to do this.  A pattern I have used with clients is to create process definition templates with expressions: where you create common process patterns that use expressions `${...}` throughout the model to essentially make it configurable at runtime. A example model could be "start form + approval step + approved or denied step". 

This way users can configure a pattern rather than inventing the pattern from scratch.  You still have the advanced ability to build from scratch, but the patterns help the typical content writers in a CMS type env compared to devs building a orchestration. 

Let me see what I can move around to make some time this week to get the micronaught stuff started to give you something to start building from.  

Are you building your UI in gentics?

Take a look at the the models here https://github.com/StephenOTT/OpenC2-Orchestrator/blob/master/README.md#openc2-orchestrator-with-worker-using-a-proxy-network-to-limit-internal-network-exposure. Should give you some ideas on possible ways to structure your workers, especially if you end up building the workers in vertx

**Comment by jdbranham** [comment]
Thank for the insight!
In-flight updates will be an issue, I'll need to manage that expectation with the client.

Good advice about the workflow templates. They're planning on hiring developers for the handover, and have asked me what skillsets they should look for - so maybe we can get ahead of this one. Still - templates are good for consistency and efficiency, even for the seasoned folks.

A web-based BPM editor is out of scope for the current phase, so we'll be using the Zeebe modeler desktop app, and have an interface to download/upload the workflows.
I was thinking we could store the BPMN in gentics, and have an eventbus listener deploy the workflow.
The rest of the UI is custom React - the gentics UI will probably be hidden from everyone but admin. 

I'll need to digest the OpenC2 references - this is new to me.
We've also implemented the Kafka-connect sink/source-worker for Zeebe -
https://github.com/zeebe-io/kafka-connect-zeebe
Trying to decouple the components more.
For example, we're feeding the gentics eventbus messages into Kafka so another service can start a workflow based on an event-to-workflow mapping.

I'll get familiar with Micronaut now - it looks interesting.

**Comment by jwulf** [comment]
A plugin to the Zeebe modeler to upload the BPM, using the zeebe-node client, would be easy to write. The main blocker for me just doing it is knowing how to add a plugin to the modeler.

**Comment by StephenOTT** [comment]
@jwulf what would the plugin solve for this context?  Good chance that you would be uploading through a proxy system to manage login, auth and authz rather than directly to the cluster. (This is the same reason for the management server deploy worker capability https://github.com/StephenOTT/Quintessential-Tasklist-Zeebe/blob/master/README.md#deploy-workflow

**Comment by StephenOTT** [comment]
@jdbranham update for you, I should have initial release with basics in place tomorrow

**Comment by jwulf** [comment]
> @jwulf what would the plugin solve for this context? Good chance that you would be uploading through a proxy system to manage login, auth and authz rather than directly to the cluster. (This is the same reason for the management server deploy worker capability https://github.com/StephenOTT/Quintessential-Tasklist-Zeebe/blob/master/README.md#deploy-workflow

In dev you can ignore that, which is sufficient for an MVP.

**Comment by StephenOTT** [comment]
little behind but conversion has started here: https://github.com/StephenOTT/Quintessential-Tasklist-Zeebe/tree/micronaut

**Comment by StephenOTT** [comment]
Update for you on migration:
1. Form server and validation logic has been converted and tested
2. User task controllers, storage and logic has been 90+% migrated. Few smaller updates around the specific logic of actions like claim/unclaim/assign. But base is there is examples of the pattern.   I also added form versioning so you can work with multiple schemas against a single form key so you can version your implementations. 
3. Zeebe manager server to deploy a workflow has been migrated
4. Zeebe start workflow has been migrate but needs more testing. 
5. Zeebe testing has been moved to TestContainers so make sure you have local docker to run your tests against the Zeebe clients. 

Just mainly cleanup and adding in the QoL adjustments.

**Comment by StephenOTT** [comment]
Update. All the parts have been added for recreation of end to end.  You can deploy workflows, start workflow instances, the worker will generate user tasks in the db, add forms and form schemas, submit for task completion using form submission data, etc 

Everything is rewritten in rxjava. 
And the worker is a custom worker wrapper using the zeebe java client. The worker is rxjava and will do batch + repoll as the pool in the batch is available. 
So example: say you have a python worker that executes scripts that take 1min to complete and you poll and have the batch max jobs set to 10: on first poll you find 1 job, and start execution, in parallel the worker will repoll for 9 jobs. 
There is the user task worker, and an example python executor worker that generate from the same GenericWorker

User task form submissions that pass form validation will have the job completed in zeebe and the user task in the db updated to a completed state. 

Round trip submission to complete a user task in dev modes of full stack on ~low-powered laptop is sub 600ms
